File: fsharp/Extended.fs
module Extended

let calculateTotalCost
    (materialCost: int -> int)
    (assemblyCost: int -> int)
    (packagingCost: int -> int)
    (quantity: int)
    : int =
    let material = materialCost quantity
    let assembly = assemblyCost quantity
    let packaging = packagingCost quantity
    material + assembly + packaging

let standardMaterialCost: int -> int =
    fun quantity -> quantity * 100

let standardAssemblyCost: int -> int =
    fun quantity -> quantity * 50

let standardPackagingCost: int -> int =
    fun quantity -> quantity * 20

let result =
    calculateTotalCost
        standardMaterialCost
        standardAssemblyCost
        standardPackagingCost
        10

File: scala/src/main/scala/operationabstraction/scalaexample/Extended.scala
package operationabstraction.scalaexample

object Extended:
  trait ManufacturingCost:
    def materialCost(quantity: Int): Int
    def assemblyCost(quantity: Int): Int

  trait PackagingCost:
    def packagingCost(quantity: Int): Int

  def calculateTotalCost(
      manufacturingCost: ManufacturingCost,
      packagingCost: PackagingCost,
      quantity: Int
  ): Int =
    val material = manufacturingCost.materialCost(quantity)
    val assembly = manufacturingCost.assemblyCost(quantity)
    val packaging = packagingCost.packagingCost(quantity)
    material + assembly + packaging

  object StandardManufacturingCost extends ManufacturingCost:
    def materialCost(quantity: Int): Int =
      quantity * 100

    def assemblyCost(quantity: Int): Int =
      quantity * 50

  object StandardPackagingCost extends PackagingCost:
    def packagingCost(quantity: Int): Int =
      quantity * 20

  def result: Int =
    calculateTotalCost(StandardManufacturingCost, StandardPackagingCost, 10)
